### PR TITLE
CUMULUS-2676: control concurrency for DiscoverGranules duplicate checking 

### PR DESCRIPTION
**Pull request by markdboyd**
Merging #2457 into `master`

### CODE REVIEW

File: docs/workflow_tasks/discover_granules.md [hunk 1]
```diff
@@ -78,6 +78,16 @@ filename occurs as described above.
 When set on the task configuration, the value applies to all collections during
 discovery.  Otherwise, this property may be set on individual collections.
 
+#### Concurrency
+
+A number property that determines the level of concurrency with which granule duplicate checks are performed when `duplicateGranuleHandling` is `skip` or `error`.
+
+Limiting concurrency helps to avoid throttling by the AWS Lambda API and helps to avoid encountering account Lambda concurrency limitations.
+
+We do not recommend increasing this value to a level at which you may encounter throttling errors, but if you are seeing Lambda.Timeout errors when discover-granules discovers a large number of granules with `skip` or `error` duplicate handling, you may either wish to increase this value or convert the discover-granules task to an ECS activity, which does not face similar runtime constraints.
```
Comment (Jkovarik) [inline docs/workflow_tasks/discover_granules.md:87]: Nit: 

This might be better said if we went with something like *emphasis on like*:  "We do not recommend increasing this value unless you are seeing Lambda.Timeout errors when running this task against a large number of granules with `skip` or `error` duplicate handling.   If you do, this value allows for tuning, but be aware that larger values increase the potential for overwhelming the AWS Lambda API for low values of `Concurrent Executions`, as each granule requires an API lambda invocation to check for granule status.

In short, I think we should put the reason for tweaking this first, with the scaling concern caveat last, and reference briefly why it's a concern.

File: tasks/discover-granules/index.js [hunk 2]
```diff
@@ -253,14 +257,19 @@ const filterDuplicates = async (granuleIds, duplicateHandling) => {
  *                                    matching files for each
  *
  * @param {string} duplicateHandling - flag that defines this function's behavior (see description)
+ * @param {number} concurrency - pMap concurrency
```
Comment (Jkovarik) [inline tasks/discover-granules/index.js:260]: we should note the answer to "concurrency for *what*" here.

File: tasks/discover-granules/index.js [hunk 3]
```diff
@@ -253,14 +257,19 @@ const filterDuplicates = async (granuleIds, duplicateHandling) => {
...
  *
  * @returns {Object} returns filesByGranuleId with applicable duplciates removed
  */
-const handleDuplicates = async (filesByGranuleId, duplicateHandling) => {
+const handleDuplicates = async (filesByGranuleId, duplicateHandling, concurrency) => {
```
Comment (Jkovarik) [inline tasks/discover-granules/index.js:264]: Nit: consider parameterizing this.

File: tasks/discover-granules/index.js [hunk 4]
```diff
@@ -231,15 +233,17 @@ const checkGranuleHasNoDuplicate = async (granuleId, duplicateHandling) => {
...
  *
  * @returns {Array.string} returns granuleIds parameter with applicable duplciates removed
  */
-const filterDuplicates = async (granuleIds, duplicateHandling) => {
-  const keysPromises = granuleIds.map((key) =>
-    checkGranuleHasNoDuplicate(key, duplicateHandling));
-
-  const filteredKeys = await Promise.all(keysPromises);
-  return filteredKeys.filter(Boolean);
+const filterDuplicates = async (granuleIds, duplicateHandling, concurrency) => {
```
Comment (Jkovarik) [inline tasks/discover-granules/index.js:240]: Nit: consider parameterizing this.

File: tasks/discover-granules/index.js [hunk 5]
```diff
@@ -231,15 +233,17 @@ const checkGranuleHasNoDuplicate = async (granuleId, duplicateHandling) => {
  *
  * @param {string[]} granuleIds - Array of granuleIds to filter
  * @param {string} duplicateHandling - flag that defines this function's behavior (see description)
+ * @param {number} concurrency - pMap concurrency
```
Comment (Jkovarik) [inline tasks/discover-granules/index.js:236]: Same comment as below re: docstring contents

**Review by Jkovarik — APPROVED** [review]
This looks good to me, nice work.    Just a couple of nits/comments for your consideration, but nothing that should hold up approval.
